mpy-cross: WASM target and Python wheel.

jimmo opened the pull request:
This PR adds:
 - A way to compile mpy-cross to a self-contained WASM (not WASI) binary. This binary exposes an entry point to take Python source as bytes (as well as the various options), and return the generated .mpy file bytes. This could be used directly from Javascript/Node (will add an example).

 - A way to use this binary from the mpy_cross Python wrapper (currently via the `wasmtime`, `pywasm3` and `wasmer` runtimes). Between these three runtimes, all of the host targets and architectures we need should be covered. The performance in both these runtimes is excellent. Initial tests show very little difference to the native binary. It also works with pywasm (which is a pure-Python WASM runtime), but too slow to be of practical use.

This is an alternative to #10834. (cc @dlech @andrewleech). Instead of having to go through the whole process to generate wheels for every architectures and platforms, we generate a single .wasm binary, and rely on there being a WASM runtime available instead for the target (which there seems to be). We can also ship multiple .wasm binaries in the (single) wheel for all the different bytecode versions we want to support.

In more detail:
* Split mpy-cross Makefile and mpconfigport.h into native and wasm versions (with shared common config).
* Implement the wasm entry point. This disables GC (to avoid needing to worry about register tracking) and NLR (not necessary for compilation).
* Makes the Python wrapper able to use different compiler backends, and implements a native and pywasm3 backend.
* Add argument parsing to the `__main__` entry point (which can now use either backend).
* Add a pyproject.toml to generate the wheel

Future work:
* Auto-select backend based on desired bytecode version.
* javascript example (i.e. browser-based mpy-cross).
* Investigate performance optimisations for pywasm.

_This work was funded through GitHub Sponsors._

jimmo committed:
py/persistentcode: Helper to get number of NLR regs for arch.

This lets mpy-cross just set the native arch without having to know
the mapping.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/nlr: Allow a port to provide a no-op NLR.

The port is required to terminate on exception, and does so by enabling
MICROPY_NLR_TERMINATE and provide an implementation of nlr_terminate.

This would only be used in a situation where you didn't expect exceptions
to be raised, or if terminating the program was a valid thing to do.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
mpy-cross: Split mpconfigport.h into common and target-specific.

This is step 1 of adding a WASM target to mpy-cross. Splits the config
between mpy-cross specific (i.e. compiler configuration) and target
specific (i.e. type config).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
mpy-cross: Add a WASM binary target.

This compiles mpy-cross.wasm, which can be run either in the browser
or via any WASM runtime.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
mpy-cross/mpy_cross: Multiple backend support for mpy-cross.

This adds support for using either a native or WASM binary in the
Python wrapper for mpy-cross.

WASM support is provided by wasmtime, pywasm3 and wasmer. It will use
whichever is available. At least one should be an option on every supported
platform/architecture.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> We've been running mpy-cross in the browser for several years now for Pybricks. For reference, here is how we are doing it: https://github.com/pybricks/pybricks-micropython/tree/master/npm/mpy-cross

I did wonder if maybe you had because I saw your commit 115acadf9211fddc8b5857af287daa622dc615c1 but wasn't sure if that was an experiment or not.

Unfortunately I don't think this approach works for WASM, only for a .js output. With a plain WASM runtime we don't have a filesystem etc. (WASI would change that potentially though).

> 
> > * NLR (not necessary for compilation)
> 
> Even if it raises a `SyntaxError` while parsing?

Yep. Functional NLR is only necessary if you want to be able to handle the exception and resume somewhere in the stack. In MicroPython, any exception in the compiler is always handled at the top level (in other words, the compiler doesn't use exception handling for program flow, only error handling). So in wasm-mpy-cross, the NLR depth was always 1 anyway, and the top level handler terminates the program, which means we can get away with just intercepting the error at the time it's raised.

So if SyntaxError is raised, the nlr_jump will call the mpycross_error extern (provided by Python) with the formatted error output, which then terminates the WASM environment.

jimmo commented:
Added `wasmer` backend.

dpgeorge commented on mpy-cross/Makefile.wasm:
```diff
@@ -0,0 +1,42 @@
+PROG ?= mpy-cross.wasm
```
A comment to describe what this file is used for and how to use it.

dpgeorge commented on mpy-cross/wasm.c:
```diff
@@ -0,0 +1,162 @@
+    MP_UNREACHABLE
+}
+
+mp_import_stat_t mp_import_stat(const char *path) {
```
Maybe it can disable `MICROPY_ENABLE_EXTERNAL_IMPORT` and then this function isn't needed (that could be done for the existing native variant as well).

dpgeorge commented on mpy-cross/wasm.c:
```diff
@@ -0,0 +1,162 @@
+
+EMSCRIPTEN_KEEPALIVE
+const uint8_t *mpycross_compile(const char *input_name, const uint8_t *input_data, size_t input_len, uint8_t default_emit_opt, uint8_t native_arch, uint8_t small_int_bits, size_t *output_len, int *result) {
+    mp_stack_set_limit(40000 * (sizeof(void *) / 4));
```
Is this needed if stack checking is disabled?

dpgeorge commented on mpy-cross/wasm.c:
```diff
@@ -0,0 +1,162 @@
+NORETURN void nlr_jump(void *val) {
+    vstr_t buf;
+    vstr_init(&buf, 1024);
+    mp_print_t fd_print = {&buf, fd_print_vstr};
```
You can use `vstr_init_print()` instead.

jimmo replied:
Neat! Done (and below).

dpgeorge commented on mpy-cross/wasm.c:
```diff
@@ -0,0 +1,162 @@
+    return MP_IMPORT_STAT_NO_EXIST;
+}
+
+STATIC int compile_and_save(const char *input_name, const uint8_t *input_data, size_t input_len, const char *source_file, vstr_t *output_data) {
```
Maybe just pass in a `mp_print_t` directly, instead of a `vstr_t`?

dpgeorge commented on mpy-cross/mpy_cross/compiler_native.py:
```diff
@@ -0,0 +1,108 @@
+    def _exec(self, args):
+        try:
+            st = os.stat(self._binary)
+            os.chmod(self._binary, st.st_mode | stat.S_IEXEC)
```
Why does it try to make it executable, surely it already is executable?

jimmo replied:
I'm not sure. This originally comes from the current PyPI package -- https://gitlab.com/alelec/mpy_cross/-/blob/master/mpy_cross/__init__.py#L16

@andrewleech do you remember the history here?

dpgeorge commented:
This looks really good!

I wonder if we should use the term "variant" here to describe the native vs wasm version of mpy-cross.  And then have a variants subdir like the unix port, that contains the various mpconfigport_xxx.h files.

jimmo committed:
py/py.mk: Add a make variable for setting MP_CONFIGFILE.

As py.mk also needs to know the path to the port config file, a port
should use this rather than setting CFLAGS directly.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented on mpy-cross/Makefile.wasm:
```diff
@@ -0,0 +1,42 @@
+CWARN += -Wextra -Wno-unused-parameter -Wpointer-arith
+CFLAGS += $(INC) $(CWARN) -std=c99 $(COPT) $(CFLAGS_EXTRA)
+CFLAGS += -fdata-sections -ffunction-sections -fno-asynchronous-unwind-tables
+
```
I don't think `MODULARIZE` has any meaning when outputting WASM? (i.e. it's a Javascript thing)

jimmo commented on mpy-cross/mpconfigport_wasm.h:
```diff
@@ -0,0 +1,44 @@
+typedef long mp_off_t;
+
+#include <alloca.h>
+#include <stdint.h>
```
This was inherited from the existing mpconfigport.h. You're right though stdint.h isn't needed.

jimmo commented on mpy-cross/wasm.c:
```diff
@@ -0,0 +1,162 @@
+#include "py/stackctrl.h"
+#include "genhdr/mpversion.h"
+
+// This must be zero, we have no way to print.
```
Yep for sure. The comment was implying "printing is not currently implemented" rather than "it is technically impossible to print". Updated the comment.

jimmo commented on py/nlr.h:
```diff
@@ -89,8 +89,12 @@
     #define MICROPY_NLR_MIPS (1)
     #define MICROPY_NLR_NUM_REGS (MICROPY_NLR_NUM_REGS_MIPS)
 #else
+    #ifndef MICROPY_NLR_SETJMP
```
The idea here is to be able to say "I don't have support for setjmp" (by setting MICROPY_NLR_SETJMP to zero), in which case, if the arch-specific configuration cannot be done (because it falls through the if/elif), then don't try and enable setjmp.

So it works because like you say, the top-level if evaluates to true, but we need the final fall through to just set MICROPY_NLR_NUM_REGS to zero to disable NLR entirely.

I guess what we really need is a way for the port to convey that NLR is best-effort (i.e. if the arch-specific implementation is unavailable, it doesn't matter, but don't use setjmp). But really the better option would be to have a way to disable NLR. So I've done that instead.

jimmo commented on mpy-cross/mpconfigport_main.h:
```diff
@@ -0,0 +1,119 @@
+
+// Adds mp_raw_code_save_file as a helper to write .mpy files to the
+// filesystem.
+#define MICROPY_PERSISTENT_CODE_SAVE_FILE (1)
```
Yeah I see what you mean. My thinking here was that mpy-cross/main.c needs `mp_raw_code_save_file` to exist, and that's what setting `MICROPY_PERSISTENT_CODE_SAVE_FILE` does -- it provides the default implementation.

But it's weird to have non-embedded OS-specific code in py/ and it doesn't really make sense for mpy-cross's mpconfigport to know about the platforms that something in py/persistentcode.c works on.

So I think it makes more sense to move `mp_raw_code_save_file` out of persistentcode.c and put it in `mpy-cross/main.c`, so I've done that instead.

jimmo commented:
> I wonder if we should use the term "variant" here to describe the native vs wasm version of mpy-cross. And then have a variants subdir like the unix port, that contains the various mpconfigport_xxx.h files.

I agree it's conceptually very similar to unix variants... although the variants have a lot more in common than native-vs-wasm for mpy-cross (i.e. they're completely independent makefiles). I'm happy to change it, but I think what we have is fine.

jimmo commented on mpy-cross/mpconfigport.h:
```diff
@@ -90,76 +76,10 @@
-
-#endif
+// Specific configuration for compiling mpy-cross as a native binary.
+#include "mpconfigport_main.h"
```
I've renamed it to "mpconfigport.h" because making the Windows Visual Studio project correctly generate the `MP_CONFIGFILE` variable is too difficult. (In summary: you can easily add the preprocessor definition, using `%22` for the quotes, but then these quotes get lost passing them down to makeqstrdata.py. You can do further transformations to re-do the quote escaping but it's a total mess and not worth the hassle). But I've kept it using MP_CONFIGFILE for the wasm build to set "mpconfigport_wasm.h".

I think conceptually it also makes sense, we have the "default" build (similar to how the Makefile has no suffix, hence it should be just "mpconfigport.h" too), and the "wasm" build is the special one. And this side-steps the awkwardness of what to call it... "main", "native", etc.

jimmo committed:
mpy-cross/pyproject.toml: Add pyproject.toml.

This will generate a wheel using the WASM binary.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
I've added `wasmtime` as a third WASM runtime, and a pyproject.toml to generate a wheel (largely copied from the mpremote pyproject.toml).

Some notes on the three runtimes:

- pywasm is distributed only as a source wheel, so it works just about everywhere, but requires the target to have a C compiler and the CPython development package (as it compiles as a Python extension).
- wasmtime provides binary packages https://pypi.org/project/wasmtime/#files for x64 Windows, Linux, macOS, as well as 64-bit aarch64 Linux (e.g. works on a Raspberry Pi 3) and M1/M2 macOS.
- wasmer provides binary packages https://pypi.org/project/wasmer/#files for x64 Windows, Linux, macOS (but only up to Python 3.10), as well as M1/M2 macOS (but only Python 3.9).

So right now I've just listed wasmtime in requirements.txt. I think with some clever tricks with https://peps.python.org/pep-0508/ we can make the requirements.txt install a suitable runtime for the current target.

dpgeorge commented on mpy-cross/library.js:
```diff
@@ -0,0 +1,30 @@
+ */
+
+mergeInto(LibraryManager.library, {
+    mpycross_error: function(buf, len) {
```
Does this need to be defined?  It's inserted into the wasm namespace by the various backends, so this here never gets called.

jimmo replied:
Yes this file is an input to emcc to tell it what to expect will be provided by the runtime. At a minimum you can write:

```js
mergeInto(LibraryManager.library, {
    mpycross_error: null
});
```

but I think it's clearer to include the signaure?

dpgeorge replied:
OK I see.

Yes better to include the signature, but add a comment that it will be replaced with a real implementation when instantiating it in a WASM runtime.

jimmo replied:
Done

dpgeorge commented on py/mpconfig.h:
```diff
@@ -1814,6 +1814,13 @@ typedef double mp_float_t;
+// A port may explicitly disable the NLR if it is unable to provide a suitable
+// implementation. If so, it must provide nlr_jump, and handle that in some
+// way (i.e. terminate).
+#ifndef MICROPY_NLR
```
Do you think disabling this will be useful for more than just a mpy-cross wasm workaround?  I preferred your previous solution, just a simple way to overridde NLR in "emergency" situations.

jimmo replied:
I don't think so (although maybe it's useful when first starting a port?).

I do like that this way is very explicit as to what it's doing -- everything is a no-op except nlr_jump.

The previous solution was a bit awkward in that it relied on the particular interpretation of disabling setjmp.

jimmo replied:
I've re-done this by adding a `MICROPY_NLR_TERMINATE` option. This is how a port (e.g. mpy-cross.wasm) signals that it explicitly wants this disable, but it's overall a much smaller change to nlr.{ch}. The idea is that it mirrors the existing `MICROPY_NLR_SETJMP` which is how a port signals that it explicitly wants the setjmp implementation.

dpgeorge commented on mpy-cross/main.c:
```diff
@@ -64,6 +68,28 @@ STATIC void stderr_print_strn(void *env, const char *str, size_t len) {
+    (void)ret;
+}
+
+void mp_raw_code_save_file(mp_compiled_module_t *cm, const char *filename) {
```
This function could be useful for more than just mpy-cross.  Eg it could be used by the unix port if it ever supports saving .mpy files, for example like `micropython.save_mpy(...)`.  So that's why I thought it was OK to keep it in `py/persistentcode.c`, similar to how `mp_lexer_new_from_fd()` is in `py/lexer.c` and guarded by `MICROPY_HELPER_LEXER_UNIX`.

dpgeorge commented on mpy-cross/pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+description = "MicroPython cross-compiler"
+readme = "README.md"
+authors = [
+    {name = "Damien George", email = "damien@micropython.org"},
```
Maybe we need something more generic here, like "The MicroPython authors"...

dlech commented on mpy-cross/requirements.txt:
```diff
@@ -0,0 +1,14 @@
+# wasmtime has the best general compatibility.
+wasmtime==11.0.0
```
It would be nice if these were optional dependencies so that we don't always have to install all runtimes, e.g. `pip install mpy-cross[pywasm3]` to get the runtime you actually want to use.

jimmo replied:
I'd like to come up with a requirements.txt that works for as many people as possible. I think what I can do is a combination of "Extras" and "Environment Markers" (following https://peps.python.org/pep-0508/) such that `pip install mpy_cross` "just" works by default, but can be customised with e.g. `pip install mpy-cross[pywasm3]` like you suggest.

At this stage, I think the best option is to come up with an environment marker that will install wasmtime by default for all the supported wasmtime targets, and then the others can be manually chosen.

jimmo replied:
I thought I understood PEP-508 enough, but it seems like I don't.

I've updated the branch with a requirements.txt that will correctly install wasmtime on the platforms that wasmtime supports (64 bit windows, linux, linux-arm, mac). But I don't know how to declare an optional dependency such that e.g. `pip install mpy-cross[pywasm3]` will do what you'd expect (i.e. install the optional dependency with the "pywasm3" extra tag).

@dlech Do you know what I'm missing here? Thanks

dlech replied:
AFAIK, extras are only additive. So the first thing that comes to mind is that we would need to split this up into an `mpy-cross-core` and an `mpy-cross` package. 

The `mpy-cross` package could include the `__main__.py` and `mpy-cross` script and the "just-works" dependencies. The dependencies could either be `mpy-cross-core[all]` or some platform-specific variations.

The `mpy-cross-core` could contain only the programmatic API and the `mpy-cross` binaries. All dependencies in this package (wasm runtimes) would be under an extra flag, i.e. consumers must either specify an extra dependency or supply their own `mpy-cross` executable.

I suppose we could take this one step further and put the binaries and backend code for each of the various backends in separate `mpy-cross-backend-xyz` packages, but the binaries are probably small enough that it isn't worth the extra effort to go this far.

dlech replied:
Thinking about it a bit more, having the suggested package split could also allow the `mpy-cross` package to follow the existing versioning scheme of exactly matching the MicroPython version while the `mpy-cross-core` package could actually use proper semver for the Python package itself.

dlech commented on py/nlr.h:
```diff
@@ -89,24 +103,26 @@
+    #else
+        // Enable the fall-back setjump implementation.
+        #define MICROPY_NLR_SETJMP (1)
+        // #warning "No native NLR support for this arch, using setjmp implementation"
```
Should we just delete this warning if we are not going to enable it?

jimmo replied:
I think so too... I'll leave it to @dpgeorge to decide.

dlech commented on py/nlr.h:
```diff
@@ -89,24 +103,26 @@
 
-typedef struct _nlr_buf_t nlr_buf_t;
-struct _nlr_buf_t {
+typedef struct _nlr_buf_t {
```
just bikeshedding, but I thought these typedefs were more readable the way they were

jimmo replied:
I figured it was improving consistency with how we were doing this elsewhere, e.g. https://github.com/micropython/micropython/blob/master/py/runtime.h#L66 and https://github.com/micropython/micropython/blob/master/py/pairheap.h#L48 (amongst others). (But I also like this way better, so ... I'm biased)

jimmo commented on mpy-cross/mpy_cross/compiler.py:
```diff
@@ -0,0 +1,204 @@
+        """
+        raise NotImplementedError()
+
+    def compile(self, src, dest=None, src_path=None, opt=0, march=None, emit=EMIT_BYTECODE):
```
The `compile` method already exists (it's used in makemanifest.py and is supported in the existing mpy_cross package on PyPI).

However, what if we allowed `src` and `dest` to be file-like objects? (In particular, I'm thinking `io.BytesIO`). Would that suit your use case?

dlech replied:
Yes, I think so. I'm assuming it could work like this...

If `src` is a string, pass it as a command line arg or if it is a file-like object, pipe it to stdin of the `mpy-cross` process. Likewise, if `dest` is a string, pass it as the `--output` argument or if it is a file-like object, pip it to stdout of the `mpy-cross` process.

jimmo commented on mpy-cross/mpy_cross/compiler.py:
```diff
@@ -0,0 +1,204 @@
+         - src:        The path to the .py file
+
+        Optional keyword arguments:
+         - dest:       The output .mpy file. Defaults to `src` (with .mpy extension)
```
Updated to "Google style" docstring comments, which is what we're using in the stub file work for the rest of the docs.

jimmo commented on mpy-cross/mpy_cross/compiler.py:
```diff
@@ -0,0 +1,204 @@
+
+        Optional keyword arguments:
+         - dest:       The output .mpy file. Defaults to `src` (with .mpy extension)
+         - src_path:   The path to embed in the .mpy file (defaults to `src`)
```
Sounds good, done.

jimmo commented on mpy-cross/mpy_cross/compiler.py:
```diff
@@ -0,0 +1,204 @@
+         - dest:       The output .mpy file. Defaults to `src` (with .mpy extension)
+         - src_path:   The path to embed in the .mpy file (defaults to `src`)
+         - opt:        Optimisation level (0-3, default 0)
+         - march:      One of the `NATIVE_ARCH_*` constants (defaults to
+                       None, which is treated as NATIVE_ARCH_NONE). Architecture to
```
Because of the silly tricks I was doing to declare the NATIVE_ARCH_* constants above, it was confusing the ruff linter. I've just made the constants explicitly, and updated it like you suggest.

jimmo commented on mpy-cross/mpy_cross/compiler_pywasm3.py:
```diff
@@ -0,0 +1,132 @@
+    emit_id,
+)
+
+try:
+    import wasm3
+except ImportError:
+    raise CompilerNotFoundError
```
We need this to work on Python 2.x... :'(

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.
